sys/net/application_layer/dhcpv6/client.c

```diff
@@ -524,6 +653,46 @@ static void _schedule_t1_t2(void)
     _schedule_t2();
 }
 
+static void _update_t1_t2(unsigned lease_t1, unsigned lease_t2)
+{
+    if ((lease_t1 != 0) && (lease_t2 != 0) &&
+        ((server.t1 == 0) || (server.t1 >= lease_t1)) &&
+        ((t2 == 0) || (t2 >= lease_t2))) {
+        server.t1 = lease_t1;
+        t2 = lease_t2;
+        _schedule_t1_t2();
+    }
+}
+
+static void _update_t2(unsigned lease_t1, unsigned lease_t2)
+{
+    if ((lease_t1 != 0) && (lease_t2 != 0) &&
+        (server.t1 > lease_t1) && (t2 > lease_t2)) {
+        server.t1 = lease_t1;
+        t2 = lease_t2;
+        _schedule_t2();
+    }
+}
+
+static void _update_addr_lease(const dhcpv6_opt_iaaddr_t *iaaddr, addr_lease_t *lease) {
+    if (iaaddr != NULL) {
+        uint32_t valid = byteorder_ntohl(iaaddr->valid);
+        uint32_t pref = byteorder_ntohl(iaaddr->pref);
+
+        lease->leased = 1U;
+        memcpy(&lease->addr, &iaaddr->addr, sizeof(ipv6_addr_t));
+        if (dhcpv6_client_add_addr(lease->parent.ia_id.info.netif,
+                                    &lease->addr) == sizeof(ipv6_addr_t)) {
+            DEBUG("IP ADDRESS successfully added!\n");
+            lease->pref_until = pref;
+            lease->valid_until = valid;
+
+            _set_event_timeout_sec(&deprecate_remove_timeout, &deprecate_remove_addrs,
+                                    pref);
+        }
```

Comment:
(Linux does this by adding the address as a `/128` prefix to the prefix list:

```console
$ ip -6 addr
[…]
    inet6 XXXX:XXXX:XXXX:1f00::93b/128 scope global dynamic noprefixroute 
       valid_lft 64336sec preferred_lft 64336sec
[…]
```

)

Reply:
This is done using `gnrc_ipv6_nib_pl_set`, right? :)

Reply:
Yepp, however, I believe you can (in this DHCPv6-specific context) just reuse `dhcpv6_client_conf_prefix()` (not sure how it will react to a `/128` prefix yet though ^^).

Reply:
Okay, great :) I will try it out :)

Reply:
I adjusted the code in 359a013c0f1d5e59d8a904bd3efb83225647c0cf

However, should there maybe also be a mechanism added for removing the prefix once the lease has expired? Or is this done automatically when the respective address is removed? If not, maybe the additional prefix configuration is actually something that should be done in separate PR.

Reply:
> However, should there maybe also be a mechanism added for removing the prefix once the lease has expired? Or is this done automatically when the respective address is removed? If not, maybe the additional prefix configuration is actually something that should be done in separate PR.

By adding the prefix to the prefix list the NIB should now take care of the valid and preferred timeouts actually. So all the timer stuff for that you added, shouldn't be necessary anymore, actually. Sorry, I did not think of that sooner.

You decide: Either revert 359a013 and do the whole deprecation/address removal thing in a separate PR untangling, or do it right away.

Reply:
Actually... this should also make the new stack-specific functions completely unnecessary oO. Address addition is covered by `dhcpv6_client_conf_prefix()` which lets instructs the NIB to deprecate and remove the address once the respective timers have passed, so

- `dhcpv6_client_add_addr`,
- `dhcpv6_client_remove_addr`, and
- `dhcpv6_client_deprecate_addr()`

do not need to be exposed anymore....

Reply:
No problem

> By adding the prefix to the prefix list the NIB should now take care of the valid and preferred timeouts actually. So all the timer stuff for that you added, shouldn't be necessary anymore, actually. Sorry, I did not think of that sooner.

No problem, the less code the better :) I am wondering however: This way we do not need to store the lifetimes in the lease struct as well, right? Hmm, I think then maybe (as this PR has already become quite long) I would open a seperate PR for this to make it clearer what is being refactored.

Reply:
> Actually... this should also make the new stack-specific functions completely unnecessary oO. Address addition is covered by `dhcpv6_client_conf_prefix()` which instructs the NIB to deprecate and remove the address once the respective timers have passed, so
> 
> * `dhcpv6_client_add_addr`,
> * `dhcpv6_client_remove_addr`, and
> * `dhcpv6_client_deprecate_addr()`
> 
> do not need to be exposed anymore....

Nice! I was thinking about these as well :)

Reply:
Then please revert 359a013 :-)

Reply:
Done :)